### ISSUE TITLE
`calc_*` test files produces different outputs for the same input

### ISSUE THREAD

**Issue by alemuller** [skvadrik/re2c#350]
The `test/calc_001.re` and `test/calc_002.re` produces different results for the same input.

```
$ ./calc001 1+1
Num
+
Num
EOF
```
```
$ ./calc002 1+1
Num
+
OOD
```

I know those are not example files. But I imagine others might use them as reference when learning and expect the generated parser to have the same behavior.

```
$ re2c --version
re2c 2.1.1
```

**Comment by skvadrik** [comment]
Hi @alemuller, thanks for reporting the problem.

I think this test is not intended to work correctly. It is an intermediate step in an old tutorial that shows how to use `YYFILL`, but it does so incorrectly. It does not explain the need for `YYMAXFILL` padding, and the next step calc003 which fixes the problem that you spotted also does not use `YYMAXFILL` --- instead it adds some constant offset 2 to `YYLIMIT` without an explanation why it's needed. The value 2 is in fact the value of `YYMAXFILL`, as can be seen by inspecting `YYFILL` invocations in the generated code. But adding offset without adding padding is dangerous and may introduce out of bounds reads (not in this case, as NULL serves as a sentinel, but generally it is misguided).

These old examples have historic value, but they should not be used to learn re2c. Fixing them is rather pointless since we already have detailed and up to date documentation on the website (handling the end of input https://re2c.org/manual/manual_c.html#handling-the-end-of-input and using `YYFILL` https://re2c.org/manual/manual_c.html#buffer-refilling).

It may be worth adding some warning to these files saying that this is not the official tutorial, or just deleting them if they confuse people (but it would be a pity since they show how re2c developed historically). Why did you come across them anyway?

**Comment by alemuller** [comment]
I see. Thank for the comprehensive answer.

> Why did you come across them anyway?

I searched for 'scan' on `*.re` files the repo. I was looking for some code that could be used as a starting point to write a [external scanner for a tree-sitter parser](https://tree-sitter.github.io/tree-sitter/creating-parsers#external-scanners).

Fell free to close the issue.

**Comment by skvadrik** [comment]
There are a few tokenizers for programming languages among re2c examples and tests, e.g. https://re2c.org/examples/c/real_world/example_cxx98.html, but if your input is already buffered or a NULL-terminated C-string, then you might need something simpler without `YYFILL`. In any case, if you have questions I'd be happy to help.

**Comment by alemuller** [comment]
Yeah. I've seem this. But tree-sitter only provide access to the buffer through a restrictive API.

Tree-sitter allows to describe tokens with basic regex on the language itself. But tokens with state, or that need to lookahead (one or more chars) shall be lexed by an external parser -- that's where I'm using re2c.

Tree-sitter's external parser API only allows to:
* move the cursor forward;
* consume whitespaces;
* mark the end of the token at *current* position (don't allow to go back)

It does allows to accepts token with length zero. When the token has zero length, the scan function is called again at same start position. I using this to fill a buffer, lex, and save the token end position to be set at next call of scan.

I described the approach I'm using at https://github.com/tree-sitter/tree-sitter/discussions/1092
